samples/bluetooth/peripheral_ots/src/main.c

```diff
@@ -134,11 +134,30 @@ static uint32_t ots_obj_read(struct bt_ots *ots, struct bt_conn *conn,
 	return len;
 }
 
+static ssize_t ots_obj_write(struct bt_ots *ots, struct bt_conn *conn,
+			     uint64_t id, const void *data, uint32_t len,
+			     uint32_t offset, uint32_t rem)
+{
+	char id_str[BT_OTS_OBJ_ID_STR_LEN];
+	uint32_t obj_index = (id % ARRAY_SIZE(objects));
+
+	bt_ots_obj_id_to_str(id, id_str, sizeof(id_str));
+
+	printk("Object with %s ID is being written\n"
+		"Offset = %d, Length = %d, Remaining= %d\n",
+		id_str, offset, len, rem);
+
+	memcpy(&objects[obj_index][offset], data, len);
+
+	return 0;
```

Comment:
Shouldn't this be `return len` now that you switched to using ssize_t?

Reply:
To be clear: if you think "success or error" makes more sense as return value than "number of bytes consumed or error" then I'm not demanding you use ssize_t. That why I said "whenever it makes sense" in my original reference to POSIX APIs.

Reply:
At this point I think it makes just as much sense to align this with the other Zephyr read/write APIs which does follow the POSIX APIs. To me it makes a whole lot of sense to have these APIs follow at least the L2CAP APIs since they are extending those.

Reply:
To be clear the current L2CAP receive API looks like below and doesn't have a `ssize_t` return type. It uses the "success or error" paradigm:
```
	/** @brief Channel recv callback
	 *
	 *  @param chan The channel receiving data.
	 *  @param buf Buffer containing incoming data.
	 *
	 *  @return 0 in case of success or negative value in case of error.
	 *  @return -EINPROGRESS in case where user has to confirm once the data
	 *                       has been processed by calling
	 *                       @ref bt_l2cap_chan_recv_complete passing back
	 *                       the buffer received with its original user_data
	 *                       which contains the number of segments/credits
	 *                       used by the packet.
	 */
	int (*recv)(struct bt_l2cap_chan *chan, struct net_buf *buf);
```
However, looks like the return handling for this would work even if we returned written length all the way back to the L2CAP layer:
```
	/* Receiving complete SDU, notify channel and reset SDU buf */
	err = chan->chan.ops->recv(&chan->chan, buf);
	if (err < 0) {
		if (err != -EINPROGRESS) {
			BT_ERR("err %d", err);
			bt_l2cap_chan_disconnect(&chan->chan);
			net_buf_unref(buf);
		}
		return;
	}
```
Although I am not sure what value this adds unless we check the return value against the actual incoming buffer length and terminate the connection on a mismatch? Or is there a more appropriate action to take?

From what I can tell, it looks like it makes sense to just keep this API as a success or error. If the application cannot handle the entire incoming buffer then it can reply with an error to terminate the channel.

Having said all of this, does it make sense to use `ssize_t` as the return type of this API? I am actually leaning towards going back to `int` to keep the API in line with the L2CAP APIs and be more clear that this is a "success or error" type API and not a "length or error" type API

Reply:
I think it makes sense to look at the OTS read/write as a sort of extension of L2CAP, and thus I think it makes sense to keep in line with the L2CAP API. 

That being said, since its a GATT service, using the same API as GATT (which follows the POSIX standard) also makes sense. 

Preferably L2CAP had used POSIX and this would be easy to conclude on :) 

I think I'm slightly leaning towards @jhedberg first suggestion and use ssize_t for read and write, and have the functions return < 0 for error or the length written/read. 

@kapi-no You were the original implementer of OTS, do you have any input?

Reply:
Aligned the types to return `ssize_t`. < 0 is interpreted as an error and in the case of write if the returned length is not the same as the requested write length an `-EIO` error is propagated to the L2CAP layer. Alternatively I could have called back the write with the rest of the write buffer in the case that all of length was not written but it didn't seem to add much value as it would be called back immediately and likely whatever the reason was for being unable to do the write would still be present.